### ISSUE TITLE
Optimizing configuration for sorting a memory mapped array

### ISSUE THREAD

**Issue by ekg** [SaschaWitt/ips4o#4]
ips4o is an amazing technical advance, and I'm really glad that you have released it as open source code.

In my work, I need to sort arrays that are far larger than main memory (~TB order). To do this, I've been using a semi-parallel MSD radix sort implementation that operates on a memory mapped array: https://github.com/ekg/bsort. ips4o is the first sorting algorithm I've found which consistently outperforms it. However, I wonder if I need to change the configuration parameters to reflect this particular high-latency context.

Do you have any suggestions for ways that I might configure ips4o for sorting memory mapped arrays?

**Comment by lorenzhs** [comment]
Have you tried a true external memory sorter like `stxxl::sort`? https://stxxl.org/ and https://github.com/stxxl/stxxl cc: @bingmann

**Comment by SaschaWitt** [comment]
As Lorenz said, your use case is external memory sorting, which ips4o isn't really designed for.

That said, you could try to increase the block size (must be a power of 2). Larger blocks result in more buffer space and fewer block swaps, which means fewer I/Os. Total memory usage of the algorithm is approximately the block size times the number of threads times 512 bytes. Choose a block size such that everything fits into main memory.

**Comment by ekg** [comment]
> Have you tried a true external memory sorter like stxxl::sort

I've tried others, but not stxxl. Thank you for bringing it to my attention. After reading [stxxl's sort algorithm design documentation](https://stxxl.org/tags/1.4.1/design_algo_sorting.html), I understand that the library began with sorting as a primary focus. I'll test it out.

> As Lorenz said, your use case is external memory sorting, which ips4o isn't really designed for.

Although it is true that ips4o is not designed for external memory sorting, I can confirm that, without any tuning it is outperforming methods that are designed specifically for this use case. Thanks for your suggestion about block size.

Some of this may relate to my use of SSD based systems, which can behave more like random access memory than spinning disks. In this context, in place sorting algorithms appear to do better in wall clock time than out of place ones. There are other in place sorts, but I've not found anything published, with workable code of equivalent API quality to ips4o.

CC @thegenemyers @richarddurbin

**Comment by ekg** [comment]
I think it's worth considering this use case. I've tested stxxl and find its parallel external memory sort to be far slower than ips4o on a memory mapped buffer _when the backing file is on an SSD_. I haven't been able to test on spinning or networked disks. I would assume that things there would bounded by the amount of the file that could be cached, and that stxxl might outperform the memory mapped sorts.

Here, I'm testing sorts of 64 bit unsigned integers using three disk backed approaches.

I'm working on a modern laptop with a pretty fast SSD that appears to achieve around 1GB/s of write:

```
cat /dev/zero | head -c $(echo '8 * 1000000000' | bc) | pv >x
7.45GiB 0:00:07 [1.05GiB/s]
```

My system has 4 vCPUs.

First I generated ~800MB of random data:

```
-> % cat /dev/urandom | head -c $(echo '8 * 100000000' | bc) >y
```

Sorting with stxxl's parallel external sort, set to use up to 5GB of RAM:

```
-> % cp y z && time ~/stxxl/build/examples/algo/sort_file sort z
[STXXL-MSG] STXXL_PARALLEL_MULTIWAY_MERGE
[STXXL-MSG] STXXL v1.4.99 (prerelease/Debug) (git b9e44f0ecba7d7111fbb33f3330c3e53f2b75236) + gnu parallel(20181206)
[STXXL-MSG] Disk '/home/erik/stxxl.tmp_file' is allocated, space: 15000 MiB, I/O implementation: syscall queue=0 devid=0
[STXXL-MSG] Sorting...
build/examples/algo/sort_file sort z  79.97s user 0.95s system 354% cpu 22.832 total
```

Sorting with ips4o in default configuration on a memory mapped buffer:

```
-> % cp y z && time ~/mmmultimap/bin/test_sort -i z -t 4
data length 100000000
starting sort
completed in 2.74887s
~/mmmultimap/bin/test_sort -i z -t 4  6.49s user 0.79s system 260% cpu 2.787 total
```

Sorting with parallel MSB binary radix sort:

```
-> % cp y z && time ~/bsort/bin/bsort -r 8 -k 8 -t 4 z
bsort -r 8 -k 8 -t 4 z  10.19s user 0.54s system 271% cpu 3.948 total
```

So it would appear that ips4o is almost an order of magnitude faster than stxxl's parallel external sort.

The specific stxxl test is in this branch: https://github.com/ekg/stxxl/tree/sort-file-test. You would build with `cmake -H. -Bbuild -DBUILD_EXAMPLES=ON && cmake --build build -- -j 4`. The ips4o sort test is https://github.com/ekg/mmmultimap/tree/ips4o-test, build with `cmake -H. -Bbuild && cmake --build build -- -j 4`. Parallel bsort can be found here: https://github.com/ekg/bsort, build with `cmake -H. -Bbuild && cmake --build build -- -j 4`.

Note that the sort result is identical for ips4o and stxxl, but bsort's MSB radix sort results in a different ordering.

**Comment by lorenzhs** [comment]
(stxxl defaults to a Debug build and your build instructions don't override that, was that an error in the posted instructions or did you maybe measure the wrong build configuration?)

**Comment by ekg** [comment]
Thanks @lorenzhs, I didn't catch that it built in the debug mode by default. This does make a difference, but ips4o is still substantially faster.

Using the correct build command: `cmake -H. -Bbuild -DBUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Release && cmake --build build -- -j 4`

In the small test (800MB), stxxl takes ~250% as long as ips4o:

```
-> % cp y z && time build/examples/algo/sort_file sort z
[STXXL-MSG] STXXL_PARALLEL_MULTIWAY_MERGE
[STXXL-MSG] STXXL v1.4.99 (prerelease/Release) (git fa4296bb3ce069fba4a59740596a546092e90d1f) + gnu parallel(20181206)
[STXXL-MSG] Disk '/home/erik/stxxl.tmp_file' is allocated, space: 15000 MiB, I/O implementation: syscall queue=0 devid=0
[STXXL-MSG] Sorting...
build/examples/algo/sort_file sort z  12.15s user 1.01s system 265% cpu 4.961 total
```

```
-> % cp y z && time ~/mmmultimap/bin/test_sort -i z -t 4
data length 100000000
starting sort
completed in 2.10949s
~/mmmultimap/bin/test_sort -i z -t 4  6.43s user 0.72s system 335% cpu 2.132 total
```

If I increase the test size to 8GB, I find that the difference decreases somewhat:

```
cat /dev/urandom | head -c $(echo '8 * 1000000000' | bc) >q
```

```
-> % cp q z && time build/examples/algo/sort_file sort z
[STXXL-MSG] STXXL_PARALLEL_MULTIWAY_MERGE
[STXXL-MSG] STXXL v1.4.99 (prerelease/Release) (git fa4296bb3ce069fba4a59740596a546092e90d1f) + gnu parallel(20181206)
[STXXL-MSG] Disk '/home/erik/stxxl.tmp_file' is allocated, space: 15000 MiB, I/O implementation: syscall queue=0 devid=0
[STXXL-MSG] Sorting...
build/examples/algo/sort_file sort z  155.55s user 10.65s system 207% cpu 1:20.10 total
```

```
-> % cp q z && time ~/mmmultimap/bin/test_sort -i z -t 4
data length 1000000000
starting sort
completed in 46.511s
~/mmmultimap/bin/test_sort -i z -t 4  69.88s user 24.61s system 202% cpu 46.660 total
```

**Comment by MichaelAxtmann** [comment]
This is an interesting result! In our paper, we analyse ips4o in the [Parallel External Memory Model](https://en.wikipedia.org/wiki/Parallel_external_memory) and show that the algorithm is asymptotically I/O efficient. The in-place property may guarantee small constant factors.

However, I'm wondering whether the first benchmark with 800MB of input is already executed externally with ips4o as well as stxxl? 

Are the benchmarks with ips4o and stxxl comparable? For example, you might have to flush the memory mapped file and the STXXL-vector completely to disk, e.g., by polluting the memory between data generation and measurement.
Note that ips4o might be slower in a more realistic scenario, e.g., when the input is already processed externally or when the output is the input of another external algorithm. STXXL provides much more than sorting data externally. STXXL comes with many features as streaming and minimization of I/O transfers when a pipeline of several external algorithms is executed.

We are very interested in your results! Could you share the benchmark code with us?

**Comment by ekg** [comment]
> This is an interesting result! In our paper, we analyse ips4o in the Parallel External Memory Model and show that the algorithm is asymptotically I/O efficient. The in-place property may guarantee small constant factors.

These constant factors may matter a lot. In my testing, which has focused on very large disk-backed arrays, in-place methods are consistently outperforming those that copy their input.

> However, I'm wondering whether the first benchmark with 800MB of input is already executed externally with ips4o as well as stxxl?

> Are the benchmarks with ips4o and stxxl comparable? For example, you might have to flush the memory mapped file and the STXXL-vector completely to disk, e.g., by polluting the memory between data generation and measurement.

I generate the data with a unix command that writes random data to a file. I make sure the file is a multiple of the integer size that I'll sort (in this case, 64 bit width integers, so 8 bytes). I've disabled the data generation capabilities and also the order checking in the codes that I'm using to make sure I'm only measuring sort time.

For each test, the input file is copied to another file and then the second file is sorted in place. I would expect this to dirty any caches related to the file. I get the same results irrespective of which algorithm I run first.

800MB is _very_ small for the use case I have, and it just serves to give a quick test of behavior. Even 8GB is pretty small, but again the goal is to get a practical sense of the behavior of the algorithms.

I am not being perfectly scientific in my testing. The aim here is mostly practical. I want to get an idea of any big differences in performance between the methods.

> Note that ips4o might be slower in a more realistic scenario, e.g., when the input is already processed externally or when the output is the input of another external algorithm. STXXL provides much more than sorting data externally. STXXL comes with many features as streaming and minimization of I/O transfers when a pipeline of several external algorithms is executed.

I agree. It is worth testing in real contexts. For me, the in-place property is important simply because of the scale that I am working at. It also minimizes a number of configuration issues with external sorts that can be hard to manage, such as how much memory and external disk backed buffer should we use, which depends on availability, the number of concurrent processes, and other factors. This helps with usability.

> We are very interested in your results! Could you share the benchmark code with us?

I'm glad it's interesting to you!

This is an important issue for my work, but I found very little information about the best way to drive external memory sorting on SSDs.

I've done things in a way that I think is reproducible. I'll collect it in a repository using submodules and send you a message.
